community/primitive-collections/src/main/java/org/neo4j/concurrent/DecayingFlags.java

```diff
@@ -0,0 +1,185 @@
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.concurrent;
+
+import java.util.Arrays;
+
+import static javax.xml.bind.DatatypeConverter.printHexBinary;
+
+/**
+ * This is a concurrent data structure used to track
+ * some set if boolean flags. Any time a flag is set to high,
+ * it'll remain high for a period of time, after which
+ * it'll fall back to low again. How long the flags stay high
+ * depend on how often they are toggled - meaning this uses
+ * both recency and frequency to determine which flags to keep
+ * high.
+ *
+ * The more often a flag is toggled high, the longer it'll
+ * take before it resets to low - if a flag gets set more
+ * often than sweep is called, it will always be high.
+ *
+ * This data structure is coordination free, but sacrifices
+ * accuracy for performance.
+ *
+ * Intended usage is that you'd have a set of keys you care
+ * about, and set a max time where you'd like to mark a key
+ * as low if that time passes; say 7 days.
+ *
+ * So, you'd set {@link #keepalive} to 7, and then you'd
+ * schedule a thread to call {@link #sweep()} once per day.
+ * Now, flags that were toggled once will be set to low again
+ * the next day, while flags that were extensively used will
+ * take up to seven days before falling back to low.
+ *
+ * Flags that are toggled at or more than once every day will
+ * always be high.
+ */
+public class DecayingFlags
+{
+    /**
+     * A flag in the set, with a unique index pointing
+     * to the bit that correlates to this flag.
+     */
+    public static class Key
+    {
+        private final int index;
+
+        public Key( int index )
+        {
+            this.index = index;
+        }
+
+        public int index()
+        {
+            return index;
+        }
+    }
+
+    /**
+     * To model the time-based "decay" of the flags,
+     * each flag is represented as an int that counts
+     * the number of times the flag has been toggled, up to
+     * {@link #keepalive}. Each time #sweep is called, all
+     * flags are decremented by 1, once a flag reaches 0 it
+     * is considered low.
+     *
+     * This way, if a flag is not "renewed", the flag will
+     * eventually fall back to low.
+     */
+    private int[] flags;
```

Comment:
I think this needs better synchronization; otherwise your sweeper thread might not see correct values.